doc: Documented pitfall with NS labels in CNPs (#26134)
cilium-team/cilium: Documentation/security/policy/kubernetes.rst

are used to create virtual clusters within a Kubernetes cluster. All Kubernetes objects
including `NetworkPolicy` and `CiliumNetworkPolicy` belong to a particular
namespace.

Known Pitfalls
~~~~~~~~~~~~~~

Depending on how a policy is defined and created, Kubernetes namespaces are automatically taken into account.

Network policies imported directly with the :ref:`api_ref` apply to all
namespaces unless a namespace selector is specified as described in
:ref:`example_cnp_ns_boundaries`.

Network policies created and imported as `CiliumNetworkPolicy` CRD and
`NetworkPolicy` apply within the namespace. In other words, the policy **only** applies
to pods within that namespace. It's possible, however, to grant access to and
from pods in other namespaces as described in :ref:`example_cnp_across_ns`.

Specifying the namespace by way of the label
``k8s:io.kubernetes.pod.namespace`` in the ``fromEndpoints`` and
``toEndpoints`` fields is supported as described in 
:ref:`example_cnp_egress_to_kube_system`.
However, Kubernetes prohibits specifying the namespace in the ``endpointSelector``,
as it would violate the namespace isolation principle of Kubernetes. The

> ``endpointSelector`` always applies to pods of the namespace 

Reviewer: ```suggestion
``endpointSelector`` always applies to pods in the namespace 
```

Author: TBH, I'm not sure about this. @zacharysarah you reviewed/rephrased it multiple times. Should it be `in` or `of`?
I think both are valid. We talk about pod **_of_** the namespace because they belong to it and at the same time they are **_in_** the namespace from a resource perspective.

Reviewer 2: @PhilipSchmid 🤔 I think `in` is clearer, but either works.

associated with the `CiliumNetworkPolicy` resource itself.

Using namespace-specific information like
``io.cilium.k8s.namespace.labels`` within a ``fromEndpoints`` or

> ``toEndpoints`` is supported only for a `CiliumClusterwideNetworkPolicy`

Reviewer: >Using namespace-specific information like
``io.cilium.k8s.namespace.labels`` within a ``fromEndpoints`` or
``toEndpoints`` is supported only for a `CiliumClusterwideNetworkPolicy`

Umm... CCNP is a cluster-wide resource - https://github.com/cilium/cilium/blob/master/pkg/k8s/apis/cilium.io/v2/ccnp_types.go#L25 so it applies to all namespaces. So what does ``io.cilium.k8s.namespace.labels`` mean in the context of a ccnp?

Author: > CCNP is a cluster-wide resource

Yes, it is. This means it can see labels configured on the Namespaces which are accessible via `io.cilium.k8s.namespace.labels` on the CiliumEndpoints.

and not a `CiliumNetworkPolicy`. The `CiliumNetworkPolicy` is a namespaced

> resource and therefore doesn't have any concept of namespaces.

Reviewer: >The `CiliumNetworkPolicy` is a namespaced
resource and therefore doesn't have any concept of namespaces.

It doesn't read well. Maybe you meant to say that the namespace specific annotations don't apply for CNPs?

Author: This statement is because of [André's initial statement](https://isovalent.slack.com/archives/C01FJ2PLRT7/p1686345345956319?thread_ts=1686344955.018409&cid=C01FJ2PLRT7).

Reviewer: Sure, but my original comment still stands. How about adding more specifics that the `io.cilium.k8s.namespace.labels` is ignored?

Author: Do you mean something like this?

Extending:
```
Using namespace-specific information like io.cilium.k8s.namespace.labels within a fromEndpoints or toEndpoints is supported only for a CiliumClusterwideNetworkPolicy and not a CiliumNetworkPolicy.
```
with this?
```
Using namespace-specific information like io.cilium.k8s.namespace.labels within a fromEndpoints or toEndpoints is supported only for a CiliumClusterwideNetworkPolicy and not a CiliumNetworkPolicy. Hence, io.cilium.k8s.namespace.labels will be ignored in CiliumNetworkPolicy resources.
```

Reviewer: Yes, that reads better. Thanks!

`Namespaces <https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/>`_
are used to create virtual clusters within a Kubernetes cluster. All Kubernetes objects
including `NetworkPolicy` and `CiliumNetworkPolicy` belong to a particular
namespace.

Known Pitfalls
~~~~~~~~~~~~~~

> When you work with namespaces within policies, be aware of some known pitfalls:
> 
> * Depending on how a policy is defined and created, Kubernetes
>   namespaces are automatically taken into account:
> 
>   * Network policies imported directly with the :ref:`api_ref` apply to all
>     namespaces unless a namespace selector is specified as described in
>     :ref:`example_cnp_ns_boundaries`.
> 
>   * Network policies created and imported as `CiliumNetworkPolicy` CRD and
>     `NetworkPolicy` apply within the namespace; in other words, the policy **only** applies
>     to pods within that namespace. It's possible, however, to grant access to and
>     from pods in other namespaces as described in :ref:`example_cnp_across_ns`.
> 
> * Specifying the namespace by way of the label
>   ``k8s:io.kubernetes.pod.namespace`` in the ``fromEndpoints`` and
>   ``toEndpoints`` fields is supported as described in :ref:`example_cnp_egress_to_kube_system`.
>   However, specifying the namespace in the ``endpointSelector`` is prohibited,
>   as it would violate the namespace isolation principle of Kubernetes. The
>   ``endpointSelector`` always applies to pods of the namespace 
>   associated with the `CiliumNetworkPolicy` resource itself.
> 
> * Using namespace-specific information like
>   ``io.cilium.k8s.namespace.labels`` within a ``fromEndpoints`` or
>   ``toEndpoints`` is supported only for a `CiliumClusterwideNetworkPolicy`
>   and not a `CiliumNetworkPolicy`. The `CiliumNetworkPolicy` is a namespaced
>   resource and therefore doesn't have any conception of namespaces.

Reviewer 2: Let's do away with the list altogether and let the section heading do some lifting. Also some typo fixes and clarity nits for active voice.
```suggestion
Depending on how a policy is defined and created, Kubernetes namespaces are automatically taken into account.

Network policies imported directly with the :ref:`api_ref` apply to all
namespaces unless a namespace selector is specified as described in
:ref:`example_cnp_ns_boundaries`.

Network policies created and imported as `CiliumNetworkPolicy` CRD and
`NetworkPolicy` apply within the namespace. In other words, the policy **only** applies
to pods within that namespace. It's possible, however, to grant access to and
from pods in other namespaces as described in :ref:`example_cnp_across_ns`.

Specifying the namespace by way of the label
``k8s:io.kubernetes.pod.namespace`` in the ``fromEndpoints`` and
``toEndpoints`` fields is supported as described in 
:ref:`example_cnp_egress_to_kube_system`.
However, Kubernetes prohibits specifying the namespace in the ``endpointSelector``,
as it would violate the namespace isolation principle of Kubernetes. The
``endpointSelector`` always applies to pods of the namespace 
associated with the `CiliumNetworkPolicy` resource itself.

Using namespace-specific information like
``io.cilium.k8s.namespace.labels`` within a ``fromEndpoints`` or
``toEndpoints`` is supported only for a `CiliumClusterwideNetworkPolicy`
and not a `CiliumNetworkPolicy`. The `CiliumNetworkPolicy` is a namespaced
resource and therefore doesn't have any concept of namespaces.
```

`Namespaces <https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/>`_
are used to create virtual clusters within a Kubernetes cluster. All Kubernetes objects
including NetworkPolicy and CiliumNetworkPolicy belong to a particular
namespace.

Known Pitfalls
~~~~~~~~~~~~~~

When you work with namespaces within policies, be aware of some known pitfalls:

> * Depending on how a policy is being defined and created, Kubernetes
>   namespaces are automatically being taken into account:
> 
>   * Network policies created and imported as `CiliumNetworkPolicy` CRD and
>     `NetworkPolicy` apply within the namespace, i.e. the policy only applies
>     to pods within that namespace. It is however possible to grant access to and
>     from pods in other namespaces as described below.
> 
>   * Network policies imported directly via the :ref:`api_ref` apply to all
>     namespaces unless a namespace selector is specified as described below.
> 
> * While specification of the namespace via the label
>   ``k8s:io.kubernetes.pod.namespace`` in the ``fromEndpoints`` and
>   ``toEndpoints`` fields is deliberately supported. Specification of the
>   namespace in the ``endpointSelector`` is prohibited as it would
>   violate the namespace isolation principle of Kubernetes. The
>   ``endpointSelector`` always applies to pods of the namespace which is
>   associated with the CiliumNetworkPolicy resource itself.
> 
> * Using namespace-specific information like
>   ``io.cilium.k8s.namespace.labels`` within a ``fromEndpoints`` or
>   ``toEndpoints`` is supported only for a CiliumClusterwideNetworkPolicy
>   and not a CiliumNetworkPolicy. The CiliumNetworkPolicy is a namespaced
>   resource and therefore doesn't have any conception of namespaces.

Reviewer 2: Edits for clarity, use simple present tense, no Latin abbreviations, use clear, plain language

It's not clear to me where the second example should be pointing. Please update the `TBD` with the text of the actual heading. Do the examples make the most sense in the current order, given that the first example that follows isn't (please verify) the first example linked?

Also, please verify that you intend the first example to point to `Example: Expose pods across namespaces`.
```suggestion
* Depending on how a policy is defined and created, Kubernetes
  namespaces are automatically taken into account:

  * Network policies created and imported as `CiliumNetworkPolicy` CRD and
    `NetworkPolicy` apply within the namespace; in other words, the policy _only_ applies
    to pods within that namespace. It's possible, however, to grant access to and
    from pods in other namespaces as described in ref:`Example: Expose pods across namespaces`.

  * Network policies imported directly with the :ref:`api_ref` apply to all
    namespaces unless a namespace selector is specified as described in :ref:`TBD`.

* Specifying the namespace by way of the label
  ``k8s:io.kubernetes.pod.namespace`` in the ``fromEndpoints`` and
  ``toEndpoints`` fields is supported. However, specifying the
  namespace in the ``endpointSelector`` is prohibited, as it would
  violate the namespace isolation principle of Kubernetes. The
  ``endpointSelector`` always applies to pods of the namespace 
  associated with the CiliumNetworkPolicy resource itself.

* Using namespace-specific information like
  ``io.cilium.k8s.namespace.labels`` within a ``fromEndpoints`` or
  ``toEndpoints`` is supported only for a CiliumClusterwideNetworkPolicy
  and not a CiliumNetworkPolicy. The CiliumNetworkPolicy is a namespaced
  resource and therefore doesn't have any conception of namespaces.
```

Author: Thanks for the additional inputs. I've applied those.

> Please update the TBD with the text of the actual heading.

Done

> Do the examples make the most sense in the current order, given that the first example that follows isn't (please verify) the first example linked?

Valid point. I've now changed the order according to the order of the examples. The last pitfall bullet point doesn't have an example.

> Also, please verify that you intend the first example to point to Example: Expose pods across namespaces.

Done

What did you mean with `_only_`? I assumed bold so that's why I changed that to `**only**`. Unfortunately, I didn't find anything regarding `_` styling inside https://docs.cilium.io/en/stable/contributing/docs/docsstyle/.

Now it looks like this:

<img width="764" alt="image" src="https://github.com/cilium/cilium/assets/5549063/74da4331-982a-4e1f-8a67-7608e89bc7ec">

	  ``endpointSelector`` always applies to pods of the namespace which is
	  associated with the CiliumNetworkPolicy resource itself.

> 	  Moreover, be aware that using namespaces-specific information like
> 	  ``io.cilium.k8s.namespace.labels`` within a ``fromEndpoints`` or
> 	  ``toEndpoints`` is only supported for a CiliumClusterwideNetworkPolicy
> 	  and not a CiliumNetworkPolicy. The CiliumNetworkPolicy is a namespaced
> 	  resource and therefore doesn't have any conception of namespaces.

Reviewer 2: ```suggestion
	  Moreover, using namespace-specific information like
	  ``io.cilium.k8s.namespace.labels`` within a ``fromEndpoints`` or
	  ``toEndpoints`` is supported only for a CiliumClusterwideNetworkPolicy
	  and not a CiliumNetworkPolicy. The CiliumNetworkPolicy is a namespaced
	  resource and therefore doesn't have any conception of namespaces.
```
